api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/RegisterBabyAction.java

```diff
@@ -150,6 +161,54 @@ Patient registerBaby(Patient mother, Concept gender, Date birthDatetime, Locatio
         return baby;
     }
 
+    Visit createVisit(Patient baby, Date startDatetime, Encounter motherEncounter) {
+        AdtService adtService = Context.getService(AdtService.class);
+        VisitService visitService = Context.getVisitService();
+        Visit visit = new Visit();
+        visit.setPatient(baby);
+        visit.setLocation(adtService.getLocationThatSupportsVisits(motherEncounter.getLocation()));
+        visit.setStartDatetime(startDatetime);
+        visit.setVisitType(visitService.getVisitTypeByUuid(PihEmrConfigConstants.VISITTYPE_CLINIC_OR_HOSPITAL_VISIT_UUID));
+
+        return visitService.saveVisit(visit);
+    }
+
+    Encounter createAdmissionEncounter(Visit visit, Patient patient, Date birthDatetime, Encounter motherEncounter) {
+        Encounter admissionEncounter = null;
+
+        EncounterService encounterService = Context.getEncounterService();
+        Encounter lastAdmissionEnc = null;
+        List<Encounter> motherEncounters =  motherEncounter.getVisit().getNonVoidedEncounters();
+        // look through mother's encounters for the most recent Admission encounter
+        for (Encounter enc : motherEncounters) {
+            if (enc.getEncounterType().getUuid().equalsIgnoreCase(PihEmrConfigConstants.ENCOUNTERTYPE_ADMISSION_UUID)) {
+                if ( lastAdmissionEnc == null ||
+                        ( (lastAdmissionEnc !=null) && lastAdmissionEnc.getEncounterDatetime().before(enc.getEncounterDatetime()) ) ) {
+                    lastAdmissionEnc = enc;
+                }
+            }
+        }
+        if (lastAdmissionEnc == null ) {
+            log.error("Mother of patient " + patient.getUuid() +" has no admission encounter, therefore we cannot create admission encounter for the baby");
+        } else {
+            admissionEncounter = new Encounter();
+            admissionEncounter.setPatient(patient);
+            admissionEncounter.setEncounterType(encounterService.getEncounterTypeByUuid(PihEmrConfigConstants.ENCOUNTERTYPE_ADMISSION_UUID));
+            admissionEncounter.setVisit(visit);
+            admissionEncounter.setEncounterDatetime(birthDatetime);
+            admissionEncounter.setLocation(lastAdmissionEnc.getLocation());
+            Set<EncounterProvider> providers =  motherEncounter.getEncounterProviders();
+            if (providers != null && !providers.isEmpty()) {
+                for (EncounterProvider provider : providers) {
+                    admissionEncounter.addProvider(provider.getEncounterRole(), provider.getProvider());
```

Comment:
It's a good point about the provider, I'm not sure if it should be to the provider on the mother encounter, let me ping Fiona.